File: labosi/ciklus3/vjezba1/zadatak2/src/hr/fer/zemris/ooup/lab2/model/plugins/Parrot.java
package hr.fer.zemris.ooup.lab2.model.plugins;

import hr.fer.zemris.ooup.lab2.model.Animal.Animal;

/**
 * Created by ivan on 5/28/15.
 */
public class Parrot extends Animal {
    private String animalName;

    public Parrot(String name) {
        animalName = name;
    }

    @Override
    public String name() {
        return animalName;
    }

    @Override
    public String greet() {
        return "Sto mu gromova!";
    }

    @Override
    public String menu() {
        return "brazilske orahe";
    }
}

File: labosi/ciklus3/vjezba1/zadatak1/parrot.c
//
// Created by ivan on 5/28/15.
//

#include <stdlib.h>
#include <string.h>


typedef char const *(*PTRFUN)();

struct Parrot {
    PTRFUN *vtable;
    char *name;
};

char const *name_fun(void *this) {
    return ((struct Parrot *) this)->name;
}

char const *greet_fun() {
    return "Sto mu gromova!";
}

char const *menu_fun() {
    return "brazilske orahe";
}


void *create(char const *name) {
    struct Parrot *animal = (struct Parrot *) malloc(sizeof(struct Parrot));

    animal->vtable = malloc(sizeof(PTRFUN) * 3);
    animal->vtable[0] = name_fun;
    animal->vtable[1] = greet_fun;
    animal->vtable[2] = menu_fun;

    size_t size = strlen(name) + 1;
    animal->name = malloc(size);
    memcpy(animal->name, name, size);

    return animal;
}
